File: claude-code-agents-typescript/src/codegen/language-router.ts
// src/codegen/language-router.ts — Language selection logic for multi-language codegen
import type { LanguageConfig, SupportedLanguage } from '../models/language.js';
import {
  isSupportedLanguage,
  languageConfigFor,
} from '../models/language.js';

// ── Route Result ────────────────────────────────────────────────
export interface RouteResult {
  readonly primaryLanguage: SupportedLanguage;
  readonly primaryConfig: LanguageConfig;
  readonly secondaryLanguages: readonly SupportedLanguage[];
  readonly secondaryConfigs: readonly LanguageConfig[];
  readonly scaffoldType: string;
  readonly rationale: string;
}

// ── Keyword Hints ───────────────────────────────────────────────
const KEYWORD_HINTS: Readonly<Record<string, SupportedLanguage>> = {
  django: 'python',
  flask: 'python',
  fastapi: 'python',
  pytorch: 'python',
  numpy: 'python',
  react: 'typescript',
  nextjs: 'typescript',
  express: 'typescript',
  nest: 'typescript',
  angular: 'typescript',
  vue: 'typescript',
  gin: 'go',
  echo: 'go',
  fiber: 'go',
  kubernetes: 'go',
  actix: 'rust',
  tokio: 'rust',
  warp: 'rust',
  spring: 'java',
  quarkus: 'java',
  ktor: 'kotlin',
  vapor: 'swift',
  swiftui: 'swift',
  aspnet: 'csharp',
  blazor: 'csharp',
  laravel: 'php',
  symfony: 'php',
  rails: 'ruby',
  sinatra: 'ruby',
  phoenix: 'elixir',
  play: 'scala',
  akka: 'scala',
};

// ── Environment Defaults ────────────────────────────────────────
const ENVIRONMENT_DEFAULTS: Readonly<Record<string, SupportedLanguage>> = {
  web: 'typescript',
  cli: 'python',
  serverless: 'typescript',
  library: 'typescript',
  'web-api': 'typescript',
  'full-stack': 'typescript',
  'sdk-wrapper': 'typescript',
  mobile: 'kotlin',
  systems: 'rust',
  data: 'python',
  ml: 'python',
};

// ── Scaffold Type Detection ─────────────────────────────────────
function detectScaffoldType(task: string, environment: string): string {
  const combined = `${task} ${environment}`.toLowerCase();
  if (combined.includes('serverless') || combined.includes('lambda'))
    return 'serverless';
  if (combined.includes('full-stack') || combined.includes('fullstack'))
    return 'full-stack';
  if (combined.includes('api') || combined.includes('server'))
    return 'web-api';
  if (combined.includes('cli') || combined.includes('command'))
    return 'cli';
  if (combined.includes('library') || combined.includes('sdk'))
    return 'library';
  if (combined.includes('wrapper')) return 'sdk-wrapper';
  return 'cli';
}

// ── Language Router ─────────────────────────────────────────────
export class LanguageRouter {
  route(
    task: string,
    environment = 'cli',
    preferredLanguages?: readonly string[],
  ): RouteResult {
    let primary: SupportedLanguage;
    const scaffoldType = detectScaffoldType(task, environment);

    if (preferredLanguages && preferredLanguages.length > 0) {
      const first = preferredLanguages[0]!.toLowerCase();
      primary = isSupportedLanguage(first) ? first : this.inferFromTask(task, environment);
    } else {
      primary = this.inferFromTask(task, environment);
    }

    const secondaries: SupportedLanguage[] = [];
    if (preferredLanguages) {
      for (const lang of preferredLanguages.slice(1)) {
        const lower = lang.toLowerCase();
        if (isSupportedLanguage(lower) && lower !== primary) {
          secondaries.push(lower);
        }
      }
    }

    return {
      primaryLanguage: primary,
      primaryConfig: languageConfigFor(primary),
      secondaryLanguages: secondaries,
      secondaryConfigs: secondaries.map(languageConfigFor),
      scaffoldType,
      rationale: `Selected ${primary} for ${environment} environment with ${scaffoldType} scaffold`,
    };
  }

  private inferFromTask(task: string, environment: string): SupportedLanguage {
    const taskLower = task.toLowerCase();

    // Check keyword hints
    for (const [keyword, lang] of Object.entries(KEYWORD_HINTS)) {
      if (taskLower.includes(keyword)) return lang;
    }

    // Check environment defaults
    const envDefault = ENVIRONMENT_DEFAULTS[environment];
    if (envDefault) return envDefault;

    return 'typescript';
  }
}

File: claude-code-agents-python/src/codegen/language_router.py
"""Route code generation tasks to appropriate languages and frameworks."""

from __future__ import annotations

import logging
from typing import Any

from ..models.language import LanguageConfig, SupportedLanguage

logger = logging.getLogger(__name__)


class LanguageRouter:
    """Routes code generation tasks to the most appropriate languages.

    Uses heuristics and optional DSPy pipeline predictions to select
    the primary and secondary languages for a given task.

    Usage:
        router = LanguageRouter()
        route = router.route("Build a REST API", environment="web")
        print(route.primary)
    """

    # Default language recommendations by environment
    ENVIRONMENT_DEFAULTS: dict[str, list[SupportedLanguage]] = {
        "web": [SupportedLanguage.TYPESCRIPT, SupportedLanguage.PYTHON, SupportedLanguage.GO],
        "cli": [SupportedLanguage.PYTHON, SupportedLanguage.GO, SupportedLanguage.RUST],
        "serverless": [SupportedLanguage.PYTHON, SupportedLanguage.TYPESCRIPT, SupportedLanguage.GO],
        "library": [SupportedLanguage.PYTHON, SupportedLanguage.TYPESCRIPT, SupportedLanguage.RUST],
        "mobile": [SupportedLanguage.SWIFT, SupportedLanguage.KOTLIN, SupportedLanguage.TYPESCRIPT],
        "data": [SupportedLanguage.PYTHON, SupportedLanguage.SCALA, SupportedLanguage.JAVA],
        "systems": [SupportedLanguage.RUST, SupportedLanguage.GO, SupportedLanguage.CSHARP],
        "scripting": [SupportedLanguage.PYTHON, SupportedLanguage.RUBY, SupportedLanguage.PHP],
    }

    # Keywords that hint at specific languages
    KEYWORD_HINTS: dict[str, SupportedLanguage] = {
        "django": SupportedLanguage.PYTHON,
        "flask": SupportedLanguage.PYTHON,
        "fastapi": SupportedLanguage.PYTHON,
        "react": SupportedLanguage.TYPESCRIPT,
        "next.js": SupportedLanguage.TYPESCRIPT,
        "nextjs": SupportedLanguage.TYPESCRIPT,
        "express": SupportedLanguage.TYPESCRIPT,
        "gin": SupportedLanguage.GO,
        "actix": SupportedLanguage.RUST,
        "tokio": SupportedLanguage.RUST,
        "spring": SupportedLanguage.JAVA,
        "ktor": SupportedLanguage.KOTLIN,
        "vapor": SupportedLanguage.SWIFT,
        "asp.net": SupportedLanguage.CSHARP,
        "blazor": SupportedLanguage.CSHARP,
        "laravel": SupportedLanguage.PHP,
        "rails": SupportedLanguage.RUBY,
        "phoenix": SupportedLanguage.ELIXIR,
        "play": SupportedLanguage.SCALA,
        "akka": SupportedLanguage.SCALA,
    }

    def __init__(self) -> None:
        """Initialize the language router."""
        self._configs: dict[SupportedLanguage, LanguageConfig] = {}

    def get_config(self, lang: SupportedLanguage) -> LanguageConfig:
        """Get or create a LanguageConfig for the given language."""
        if lang not in self._configs:
            self._configs[lang] = LanguageConfig.for_language(lang)
        return self._configs[lang]

    def route(
        self,
        task_description: str,
        environment: str = "cli",
        preferred_languages: list[str] | None = None,
        constraints: str = "",
    ) -> RouteResult:
        """Route a task to the appropriate language(s).

        Args:
            task_description: Description of the code to generate.
            environment: Target environment (web, cli, serverless, etc.).
            preferred_languages: User-preferred languages, or None for auto.
            constraints: Additional constraints.

        Returns:
            A RouteResult with primary language, secondaries, and configs.
        """
        # Check for explicit preferences first
        if preferred_languages:
            primary = self._resolve_language(preferred_languages[0])
            secondaries = [
                self._resolve_language(lang)
                for lang in preferred_languages[1:]
                if self._resolve_language(lang) is not None
            ]
            if primary:
                return RouteResult(
                    primary=primary,
                    secondaries=[s for s in secondaries if s is not None],
                    configs={
                        lang: self.get_config(lang)
                        for lang in [primary] + [s for s in secondaries if s is not None]
                    },
                    rationale=f"User preference: {preferred_languages}",
                )

        # Check for keyword hints in the task description
        task_lower = task_description.lower()
        for keyword, lang in self.KEYWORD_HINTS.items():
            if keyword in task_lower:
                logger.info("Keyword '%s' matched -> %s", keyword, lang.value)
                return RouteResult(
                    primary=lang,
                    secondaries=[],
                    configs={lang: self.get_config(lang)},
                    rationale=f"Keyword match: '{keyword}' -> {lang.value}",
                )

        # Fall back to environment defaults
        env_key = environment.lower().replace("-", "").replace("_", "")
        defaults = self.ENVIRONMENT_DEFAULTS.get(env_key, [SupportedLanguage.PYTHON])
        primary = defaults[0]
        secondaries = defaults[1:] if len(defaults) > 1 else []

        return RouteResult(
            primary=primary,
            secondaries=secondaries,
            configs={lang: self.get_config(lang) for lang in [primary] + secondaries},
            rationale=f"Environment default for '{environment}': {primary.value}",
        )

    def _resolve_language(self, name: str) -> SupportedLanguage | None:
        """Resolve a language name string to a SupportedLanguage enum."""
        name_lower = name.strip().lower()
        for lang in SupportedLanguage:
            if lang.value == name_lower:
                return lang
        # Common aliases
        aliases: dict[str, SupportedLanguage] = {
            "ts": SupportedLanguage.TYPESCRIPT,
            "js": SupportedLanguage.TYPESCRIPT,
            "javascript": SupportedLanguage.TYPESCRIPT,
            "py": SupportedLanguage.PYTHON,
            "rs": SupportedLanguage.RUST,
            "cs": SupportedLanguage.CSHARP,
            "c#": SupportedLanguage.CSHARP,
            "rb": SupportedLanguage.RUBY,
            "ex": SupportedLanguage.ELIXIR,
            "kt": SupportedLanguage.KOTLIN,
        }
        return aliases.get(name_lower)


class RouteResult:
    """Result of routing a task to languages."""

    def __init__(
        self,
        primary: SupportedLanguage,
        secondaries: list[SupportedLanguage],
        configs: dict[SupportedLanguage, LanguageConfig],
        rationale: str = "",
    ) -> None:
        self.primary = primary
        self.secondaries = secondaries
        self.configs = configs
        self.rationale = rationale

    @property
    def all_languages(self) -> list[SupportedLanguage]:
        """All languages (primary + secondaries)."""
        return [self.primary] + self.secondaries

    @property
    def primary_config(self) -> LanguageConfig:
        """Config for the primary language."""
        return self.configs[self.primary]

    def to_dict(self) -> dict[str, Any]:
        """Serialize to a dictionary."""
        return {
            "primary": self.primary.value,
            "secondaries": [s.value for s in self.secondaries],
            "rationale": self.rationale,
        }
